### ISSUE TITLE
map_flatten does not cover Result.map

### ISSUE THREAD

**Issue by Dionysusnu** [rust-lang/rust-clippy#7496]
Lint name: `map_flatten`

### I tried this code:
```rust
let a: Result<&str, ParseIntError> = Ok("1");
let c: i32 = a.map(|b| b.parse()).flatten().unwrap();
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3c1eced0e47dac7f563945134ddc3bc8
### I expected to see this happen:
This should fall under lint rule `map_flatten`.
The lint should advise the user to use `andThen`, instead of the `.map(..).flatten()`

### Instead, this happened:
Result is not currently covered by the lint rule.
However, it is nearly the same as Option's map_flatten, and therefore makes sense to be included.

### Meta

- `cargo clippy -V`: e.g. clippy 0.1.53 (53cb7b09 2021-06-17)
-- Although I did run `rustup update`, I'm not sure if this is the latest dev version, but the lint rule doesn't seem to have changed in any recent versions.
- `rustc -Vv`:
```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-pc-windows-msvc
release: 1.53.0
LLVM version: 12.0.1
```

**Comment by Dionysusnu** [comment]
First issue here, still relatively new to Rust. Sorry if this is bad/wrong, please tell me if so.

**Comment by xFrednet** [comment]
This is the right place to report this. It was probably not implemented as `flatten` for results is still behind a feature flag. I've created a second playground that doesn't use the `parse` method: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cd7b9051f684341f167c3aa555b1273a). Thank you for providing an example, that was very helpful! :upside_down_face: 

The implementation has to check that the wrapped error types are the same and that the feature `result_flattening` is enabled.

**Comment by camsteffen** [comment]
> The implementation has to check that the wrapped error types are the same

We already know the error types are the same if `flatten` compiled!

I think this lint should also cover `option.map().flatten()` and `iterator.map().flatten()`.
